rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java

```diff
@@ -500,6 +515,64 @@ public CompletableFuture<SemanticTokens> semanticTokensRange(SemanticTokensRange
             , Collections::emptyList);
     }
 
+    @Override
+    public CompletableFuture<List<Either<Command, CodeAction>>> codeAction(CodeActionParams params) {
+        logger.debug("codeActions: {}", params);
+
+        final ILanguageContributions contribs = contributions(params.getTextDocument());
+        final var loc = Locations.toLoc(params.getTextDocument());
+        final var start = params.getRange().getStart();
+        final var startLine = start.getLine() + 1;
+        final var startColumn = columns.get(loc).translateInverseColumn(startLine, start.getCharacter(), false);
+        
+        if (contribs != null) {
+            // first we filter out the "fixes" that were optionally sent along with earlier diagnostics
+            // and which came back with the codeAction's list of relevant (in scope) diagnostics:
+            Stream<IValue> quickfixes = params.getContext().getDiagnostics()
```

Comment:
done. I thought btw that the streams were already lazy, so all computation effectively was already on the back ground tasks that were iterating through the streams. The code outside the background task was only building the streams. But maybe I was misunderstanding this.

Reply:
Streams  are lazy, but their execution is done at collect time, so if the collect is not explicitly scheduled on the background, it will happen in the foreground. So in this specific case, it looks like all the calls scheduled here would indeed be executed inside the completable futre.

In general the pattern in this class is: do as much as you can inside of a CompletableFuture, just so that 

1. exceptions are properly translated to the server responds
2. you don't have to look at all your down stream calls if they might do a complex calculation that would block the server handler longer than needed.

Reply:
Ok thanks!